pkg/apis/certificates/v1beta1/defaults.go

```diff
@@ -17,15 +17,121 @@ limitations under the License.
 package v1beta1
 
 import (
+	"crypto/x509"
+	"reflect"
+	"strings"
+
 	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
 	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
 )
 
 func addDefaultingFuncs(scheme *runtime.Scheme) error {
 	return RegisterDefaults(scheme)
 }
+
 func SetDefaults_CertificateSigningRequestSpec(obj *certificatesv1beta1.CertificateSigningRequestSpec) {
 	if obj.Usages == nil {
 		obj.Usages = []certificatesv1beta1.KeyUsage{certificatesv1beta1.UsageDigitalSignature, certificatesv1beta1.UsageKeyEncipherment}
 	}
+
+	if obj.SignerName == nil {
+		signerName := defaultSignerNameFromSpec(obj)
+		obj.SignerName = &signerName
+	}
+}
+
+// defaultSignerNameFromSpec will determine the signerName that should be set
+// by attempting to inspect the 'request' content and the spec options.
+func defaultSignerNameFromSpec(obj *certificatesv1beta1.CertificateSigningRequestSpec) string {
+	csr, err := ParseCSR(&certificatesv1beta1.CertificateSigningRequest{Spec: *obj})
+	if err != nil {
+		// This case shouldn't be reached as we already have validation in
+		// admission that the 'request' data is a valid CSR.
+		utilruntime.HandleError(err)
+		// Set the signerName to 'legacy-unknown' as the CSR could not be
+		// recognised.
+		return certificatesv1beta1.LegacyUnknownSignerName
+	}
+	switch {
+	case isKubeletClientCSR(obj, csr):
+		return certificatesv1beta1.KubeAPIServerClientKubeletSignerName
+	case isKubeletServingCSR(obj, csr):
+		return certificatesv1beta1.KubeletServingSignerName
+	}
+	return certificatesv1beta1.LegacyUnknownSignerName
+}
+
+func isKubeletServingCSR(spec *certificatesv1beta1.CertificateSigningRequestSpec, req *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{"system:nodes"}, req.Subject.Organization) {
+		return false
+	}
+
+	// at least one of dnsNames or ipAddresses must be specified
+	if len(req.DNSNames) == 0 && len(req.IPAddresses) == 0 {
+		return false
+	}
+
+	if len(req.EmailAddresses) > 0 || len(req.URIs) > 0 {
+		return false
+	}
+
+	requiredUsages := []certificatesv1beta1.KeyUsage{
+		certificatesv1beta1.UsageDigitalSignature,
+		certificatesv1beta1.UsageKeyEncipherment,
+		certificatesv1beta1.UsageServerAuth,
+	}
+	if !equalUnsorted(requiredUsages, spec.Usages) {
+		return false
+	}
+
+	if !strings.HasPrefix(req.Subject.CommonName, "system:node:") {
+		return false
+	}
+
+	if spec.Username != req.Subject.CommonName {
+		return false
+	}
+
+	return true
+}
+
+func isKubeletClientCSR(spec *certificatesv1beta1.CertificateSigningRequestSpec, req *x509.CertificateRequest) bool {
```

Comment:
I would expect kube-controller-manager to use this same logic to determine whether it should approve and sign a certificate that requests the apiserver client signerName. You should only need `req *x509.CertificateRequest` as a parameter, and I would suggest moving it to `pkg/apis/certificates/helpers.go` and exporting it

Reply:
It also needs to accept a list of usages, which requires passing a versioned var - but I could keep it in this package and change the function to accept `(req *x509.CertificateRequest, usages []capi.KeyUsage)` then export it?

Reply:
ah, right... usages